### ISSUE TITLE
The poc is not working. 

### ISSUE THREAD

**Issue by sun-ruijiang** [chenyueqi/hotBPF#3]
I can’t verify the effect of hotbpf even though I have completely followed Readme to install the hardened kernel. The poc given by testcase cannot be executed?

**Comment by chenyueqi** [comment]
The poc.c should first be compiled into executable files

**Comment by sun-ruijiang** [comment]
Yes, I have compiled poc into an executable file and run it. 
```
make
sudo ./poc
```
The poc is always in the loop () function without causing system crash and other consequences

**Comment by sun-ruijiang** [comment]
@brant-ruan Hello, did you succeed later? I had a problem with poc operation.

**Comment by chenyueqi** [comment]
If you want to see how PoC makes an effect, you can compile the kernel with sanitizer enabled.

**Comment by sun-ruijiang** [comment]
I don't quite understand, "sanitizer enabled" means to turn on some kernel configuration options？

**Comment by sun-ruijiang** [comment]
I also want to know in the dynamic isolation section, if I want to run hotbpf, do I have to compile and run the hardened kernel? This doesn't seem to satisfy the immediate protection in the video you explained on youtube.

**Comment by chenyueqi** [comment]
Yes. sanitizer enabled means you should at least enable CONFIG_KASAN.
> I don't quite understand, "sanitizer enabled" means to turn on some kernel configuration options？

Yes, you need to compile a hardened kernel with the necessary helper functions and allocator supports. By immediate protection, it means with this customized kernel, you can enable isolation on-the-fly. If you run a vanilla kernel, the protection is just not there.
> I also want to know in the dynamic isolation section, if I want to run hotbpf, do I have to compile and run the hardened kernel? This doesn't seem to satisfy the immediate protection in the video you explained on youtube.

**Comment by sun-ruijiang** [comment]
Thank you for your guidance. 

I understand that the hardened kernel can run the hotbpf corresponding to the vulnerability for isolation and protection. If the hotbpf is turned off, it can be successfully attacked by poc, right?

Another problem is that there are only two versions of the hardened kernel. Some reports in the testcases are all vulnerabilities in other versions of the kernel. If i want to do tests, can i do them on these two hardened kernels? Is it possible that the tests are ineffective because of different kernel versions?

**Comment by chenyueqi** [comment]
The PoC demonstrates how to trigger the vulnerability without performing exploitation. 
> I understand that the hardened kernel can run the hotbpf corresponding to the vulnerability for isolation and protection. If the hotbpf is turned off, it can be successfully attacked by poc, right?

This repo is a public one of this project. As we are seeking to commercialize the project, only two versions are provided. You can migrate the code across versions. 
> Another problem is that there are only two versions of the hardened kernel. Some reports in the testcases are all vulnerabilities in other versions of the kernel. If i want to do tests, can i do them on these two hardened kernels? Is it possible that the tests are ineffective because of different kernel versions?

**Comment by sun-ruijiang** [comment]
Sorry, I have tried to recompile the kernel with sanitizer enabled, but poc still doesn't work. I can't see the effect of hotbpf. 
Does the configuration file of the hardened kernel need to be compared with the configuration file in the report and then modified? Do I have to do something else?

**Comment by chenyueqi** [comment]
The PoC was generated by Syzkaller. There would be many reasons for failing to reproduce. Configuration is one of them.

**Comment by sun-ruijiang** [comment]
I don't know what to do to see the effect of hotbpf. 
Take demo as an example, I want to know how you did it at that time, such as how to write the configuration file and choose the kernel version. 
Sorry to bother you.

**Comment by sun-ruijiang** [comment]
If you have any other suggestions later, you can reply to me at any time.

**Comment by chenyueqi** [comment]
Hi, I cannot do this debugging for you. I just updated all BPF programs to be aligned with v5.15. If you want to test the effect of hotBPF. You can first migrate the vulnerability, say `bug-kobject_add_interna`, to v5.15, and enable sanitizers to first test its existence before you use hotBPF.

**Comment by sun-ruijiang** [comment]
Thank you for your efforts, but I have a few problems. 
The first is the problem with compiling the kernel.
```
kernel/trace/bpf_trace.c:760:32: error: ‘MEM_RDONLY’ undeclared here (not in a function); did you mean ‘SHM_RDONLY’?
  760 |  .arg2_type = ARG_PTR_TO_MEM | MEM_RDONLY,
      |                                ^~~~~~~~~~
      |                                SHM_RDONLY
  CC      arch/x86/kernel/cpu/microcode/amd.o
kernel/trace/bpf_trace.c:863:37: error: ‘PTR_MAYBE_NULL’ undeclared here (not in a function)
  863 |  .arg4_type      = ARG_PTR_TO_MEM | PTR_MAYBE_NULL | MEM_RDONLY,
      |                                     ^~~~~~~~~~~~~~
```
The second is the configuration file problem. 
If I use the `config` file in `/poc/bug-kobject_add_interna` to compile the kernel, it will cause the kernel to fail to open, which is exactly what I didn't solve before, so I can't even reproduce the kernel crash reported by syzbot.
```
Gave up waiting for root file system device. Common problems:
 - Boot args (cat /proc/cmdline)
   - Check rootdelay= (did the system wait long enough?)
 - Missing modules (cat /proc/modules; ls /dev)
ALERT! UUID=<uuid> does not exist. Dropping to a shell!

BusyBox v1.30. 1 (Ubuntu 1: 1.30.1-4ubuntu6.4) built-in shell (ash)
Enter 'help' for a list of built-in commands.

(initramfs)
```

**Comment by sun-ruijiang** [comment]
Sry, I didn't know I could send pictures. This is a screenshot of the second question.
![QQ图片20230429172935](https://user-images.githubusercontent.com/67152344/235295880-603f49b5-ae71-470d-912e-0efecc82e019.png)

**Comment by chenyueqi** [comment]
As I said, I cannot debug for you. As to your questions
1. `MEM_RDONLY ` is a composable bpf type that is introduced in v5.15.37. 
2. The configuration file of `bug-kobject_add_interna` is for v5.8 as you can see at the very beginning of the file. For vulnerability migration, this file is for your reference. You cannot directly copy and paste because the kernel changes a lot between v5.8 and v5.15.

**Comment by lazyman0001** [comment]
I have successfully run poc, but hotbpf cannot run successfully.
```
bpf_load_program() err=22
0: (bf) r6 = r1
1: (85) call bpf_ktime_get_ns#5
2: (b7) r1 = 0
3: (7b) *(u64 *)(r10 -8) = r1
last_idx 3 first_idx 0
regs=2 stack=0 before 2: (b7) r1 = 0
4: (7b) *(u64 *)(r10 -16) = r1
5: (79) r1 = *(u64 *)(r6 +8)
6: (55) if r1 != 0x81b88911 goto pc+63
 R0_w=inv(id=0) R1_w=inv-2118612719 R6_w=ctx(id=0,off=0,imm=0) R10=fp0 fp-8_w=00
7: (79) r1 = *(u64 *)(r6 +24)
8: (7b) *(u64 *)(r10 -16) = r1
9: (85) call bpf_vmalloc#136
unknown func bpf_vmalloc#136
processed 10 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_st0
Error: load_bpf_file failed
```

**Comment by chenyueqi** [comment]
> 

`bpf_vmalloc` is no longer there. The design now has been changed to dedicated regions in direct mapping for DMA object isolation.
